kyuubi-zookeeper/src/main/scala/org/apache/kyuubi/zookeeper/EmbeddedZookeeper.scala

```diff
@@ -17,77 +17,83 @@
     val clientPort = normalizePort(conf.get(ZK_CLIENT_PORT))
-    val electionPort = normalizePort(conf.get(ZK_ELECTION_PORT))
-    val quorumPort = normalizePort(conf.get(ZK_QUORUM_PORT))
-    val serverId = conf.get(ZK_SERVER_ID)
     val tickTime = conf.get(ZK_TICK_TIME)
     val maxClientCnxns = conf.get(ZK_MAX_CLIENT_CONNECTIONS)
-    // TODO: Is it right in prod?
-    val deleteDataDirectoryOnClose = true
-    val customProperties =
-      (conf.get(ZK_MIN_SESSION_TIMEOUT).map { timeout =>
-        "minSessionTimeout" -> Integer.valueOf(timeout)
-      } ++ conf.get(ZK_MAX_SESSION_TIMEOUT).map { timeout =>
-        "maxSessionTimeout" -> Integer.valueOf(timeout)
-      }).toMap[String, Object].asJava
-
+    val minSessionTimeout = conf.get(ZK_MIN_SESSION_TIMEOUT)
+    val maxSessionTimeout = conf.get(ZK_MAX_SESSION_TIMEOUT)
     val hostname = conf.get(ZK_CLIENT_PORT_ADDRESS).map(InetAddress.getByName)
       .getOrElse(findLocalInetAddress).getCanonicalHostName
-    spec = new InstanceSpec(
-      dataDirectory,
-      clientPort,
-      electionPort,
-      quorumPort,
-      deleteDataDirectoryOnClose,
-      serverId,
-      tickTime,
-      maxClientCnxns,
-      customProperties,
-      hostname)
-    server = new TestingServer(spec, false)
+
+    zks = new ZooKeeperServer(dataDirectory, dataDirectory, tickTime)
+    zks.setMinSessionTimeout(minSessionTimeout)
+    zks.setMaxSessionTimeout(maxSessionTimeout)
+
+    serverFactory = new NIOServerCnxnFactory
+    serverFactory.configure(new InetSocketAddress(hostname, clientPort), maxClientCnxns)
+
     super.initialize(conf)
   }
 
   override def start(): Unit = synchronized {
-    server.start()
+    serverFactory.startup(zks)
     info(s"$getName is started at $getConnectString")
     // Stop the EmbeddedZookeeper after the Kyuubi server stopped
-    addShutdownHook(() => server.close(), SERVER_SHUTDOWN_PRIORITY - 1)
+    addShutdownHook(() => stop(), SERVER_SHUTDOWN_PRIORITY - 1)
     super.start()
   }
 
   override def stop(): Unit = synchronized {
     if (getServiceState == ServiceState.STARTED) {
-      server.close()
+      if (null != serverFactory) serverFactory.shutdown()
+      if (null != zks) zks.shutdown()
+      if (deleteDataDirectoryOnClose) deleteDirectoryRecursively(dataDirectory)
     }
     super.stop()
   }
 
   def getConnectString: String = synchronized {
-    assert(spec != null, s"$getName is in $getServiceState")
-    spec.getConnectString
+    assert(zks != null, s"$getName is in $getServiceState")
+    s"${serverFactory.getLocalAddress.getHostName}:${serverFactory.getLocalPort}"
+  }
+
+  private def normalizePort(port: Int): Int = if (port <= 0) getRandomAvailablePort else port
+
+  /**
+   * Returns with a random open port can be used to set as server port for ZooKeeper.
+   *
+   * @return a random open port or 0 (in case of error)
+   */
+  private def getRandomAvailablePort: Int = {
```

Comment:
why do we need this hack?

Reply:
When the client port is -1, the `TestingServer` will randomly pick free local port to use, but `ZooKeeperServer` does not support this.
Also refer to hadoop `MicroZookeeperService.java`.

Reply:
why is this related to zk? InetSocketAddress is already supporting specify random port

Reply:
My wrong, is `NIOServerCnxnFactory`. Wait a moment, will look into it.

Reply:
Got, we can remove `normalizePort` and `getRandomAvailablePort`.